### PR TITLE
fix(types): tools export types added

### LINKED ISSUE #2771: Type error, type files are missing while compiling
While installing editorjs in angular 18 version it is showing the type error for importing few files
`✘ [ERROR] TS2307: Cannot find module '../../src/components/block' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/api/blocks.d.ts:1:18:
      1 │ import Block from '../../src/components/block';`
`✘ [ERROR] TS2307: Cannot find module '../../src/components/tools/block' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/api/tools.d.ts:1:22:
      1 │ import BlockTool from "../../src/components/tools/block";
        ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../../src/components/utils/popover' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/configs/index.d.ts:8:14:
      8 │ export * from '../../src/components/utils/popover';
        ╵               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../src/components/utils/popover' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/index.d.ts:86:7:
      86 │ } from '../src/components/utils/popover';
         ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemDefaultBaseParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:9:
      1 │ import { PopoverItemDefaultBaseParams, PopoverItemHtmlParams, Popov...
        ╵          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemHtmlParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:39:
      1 │ ...temDefaultBaseParams, PopoverItemHtmlParams, PopoverItemParams, ...
        ╵                          ~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:62:
      1 │ ... PopoverItemHtmlParams, PopoverItemParams, PopoverItemSeparatorP...
        ╵                            ~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemSeparatorParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:81:
      1 │ ...PopoverItemParams, PopoverItemSeparatorParams, WithChildren } fr...
        ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'WithChildren'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:109:
      1 │ ...rams, PopoverItemSeparatorParams, WithChildren } from "../configs";
        ╵                                      ~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../../src/components/block' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/api/blocks.d.ts:1:18:
      1 │ import Block from '../../src/components/block';
        ╵                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../../src/components/tools/block' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/api/tools.d.ts:1:22:
      1 │ import BlockTool from "../../src/components/tools/block";
        ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../../src/components/utils/popover' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/configs/index.d.ts:8:14:
      8 │ export * from '../../src/components/utils/popover';
        ╵               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2307: Cannot find module '../src/components/utils/popover' or its corresponding type declarations. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/index.d.ts:86:7:
      86 │ } from '../src/components/utils/popover';
         ╵        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemDefaultBaseParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:9:
      1 │ import { PopoverItemDefaultBaseParams, PopoverItemHtmlParams, Popov...
        ╵          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemHtmlParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:39:
      1 │ ...temDefaultBaseParams, PopoverItemHtmlParams, PopoverItemParams, ...
        ╵                          ~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:62:
      1 │ ... PopoverItemHtmlParams, PopoverItemParams, PopoverItemSeparatorP...
        ╵                            ~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'PopoverItemSeparatorParams'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:81:
      1 │ ...PopoverItemParams, PopoverItemSeparatorParams, WithChildren } fr...
        ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~

✘ [ERROR] TS2305: Module '"../configs"' has no exported member 'WithChildren'. [plugin angular-compiler]

    node_modules/.pnpm/@editorjs+editorjs@2.30.1/node_modules/@editorjs/editorjs/types/tools/menu-config.d.ts:1:109:
      1 │ ...rams, PopoverItemSeparatorParams, WithChildren } from "../configs";`
      
      
Steps to reproduce:
1. Use angular 18 version
2. npm i @editorjs/editorjs

Device, Browser, OS: Macbook m3 pro

Editor.js version: 2.30.1

### PR DESCRIPTION
**Pull request by neSpecc**
## Problem

Some compilers can't build projects with current editor.js import because of errors like that

```
node_modules/.pnpm/@editorjs+editorjs@2.30.0/node_modules/@editorjs/editorjs/types/api/tools.d.ts:1:22:
  1 │ import BlockTool from "../../src/components/tools/block";
```

Resolves #2771

## Cause

In #2718 we introduced `api.tool.getBlockTools()` method that returns `BlockTool[]` objects which declared under the npmignored `/src` folder. 

### Confusing names 

We have a confusing names in `src/components/tools/`:

- `BlockTool` — it is not a block tool (see [types/tools/block-tool.d.ts](https://github.com/codex-team/editor.js/blob/next/types/tools/block-tool.d.ts)). It is more like block tool adapater
- same with `InlineTool` and `BlockTune`

So I've renamed them to `BlockToolAdapter`, `InlineToolAdapter`, `BlockTuneAdapter` across all the code.

Now BlockTool has only one meaning declared in [types/tools/block-tool.d.ts](https://github.com/codex-team/editor.js/blob/next/types/tools/block-tool.d.ts)

### Types for adapters

I've manually created types for adapters:

```
/types/tools/adapters/block-tool-adapter.d.ts
/types/tools/adapters/inline-tool-adapter.d.ts
/types/tools/adapters/block-tune-adapter.d.ts
```

Classes in `src/components/tools/` implements those adapters interfaces.

### Also

- Enums are moved to `/types` as well, but under `.ts` files instead of `d.ts` (complier can't find enums values in d.ts)
- Popover types are moved to `/types` as well
- Added `@/types` alias
- Eslint now enforces to use "import type" when only type is needed.

We manually maintain `/types` which may be not-best solution today. We can consider autogenerating types in a future.

### CODE REVIEW

File: types/tools/wrappers/base-tool-wrapper.d.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { Tool, ToolConstructable, ToolSettings } from '..';
+import type { SanitizerConfig, API as ApiMethods } from '../..';
+
+import { ToolType } from './tool-type';
+import { InlineToolWrapper } from './inline-tool-wrapper';
+import { BlockToolWrapper } from './block-tool-wrapper';
+import { BlockTuneWrapper } from './block-tune-wrapper';
+
+export interface BaseToolWrapper<Type extends ToolType, ToolClass extends Tool> {
```
Comment (gohabereg) [inline types/tools/wrappers/base-tool-wrapper.d.ts]: Can you describe what this interface represents and why do we need to wrap tools

Comment (gohabereg) [reply types/tools/wrappers/base-tool-wrapper.d.ts]: Also let's maybe call it a factory or builder, the wrapper is not really clear

Comment (neSpecc) [reply types/tools/wrappers/base-tool-wrapper.d.ts]: it is the question for you. It is just the type describing your BlockTool etc classes (which are actually wrappers around tools)

**Review by gohabereg — APPROVED** [review]

File: src/components/block/index.ts [hunk 2]
```diff
@@ -9,24 +9,25 @@ import {
...
 
-import BlockTune from '../tools/tune';
-import { BlockTuneData } from '../../../types/block-tunes/block-tune-data';
-import ToolsCollection from '../tools/collection';
+import type BlockTune from '../tools/tune';
+import type { BlockTuneData } from '../../../types/block-tunes/block-tune-data';
+import type ToolsCollection from '../tools/collection';
 import EventsDispatcher from '../utils/events';
-import { TunesMenuConfigItem } from '../../../types/tools';
+import type { TunesMenuConfigItem } from '../../../types/tools';
```
Comment (TatianaFomina) [inline src/components/block/index.ts:24]: why not use new `@/types/` alias everywhere?

File: src/components/tools/base.ts [hunk 3]
```diff
@@ -1,28 +1,11 @@
...
-   */
-  Inline,
-
-  /**
-   * Block tune
-   */
-  Tune,
-}
+import { ToolType } from '@/types/tools/wrappers/tool-type';
+import type { BaseToolFactory } from '@/types/tools/wrappers/base-tool-factory';
```
Comment (TatianaFomina) [inline src/components/tools/base.ts]:
```suggestion
import type { BaseToolFactory } from '@/types/tools/factories/base-tool-factory';
```

File: src/components/block/index.ts (hunk 2 above)
Comment (neSpecc) [reply src/components/block/index.ts:24]: It is out of scope. 
I changed only places I touched. There are too many files in the pr already.

**Review by TatianaFomina — APPROVED** [review]

**Review by gohabereg — APPROVED** [review]

**Review by ilyamore88 — APPROVED** [review]

**Comment by golodnoy** [comment]
HI @neSpecc 
There are still bugs in the new version of the library. Please fix them

**Pull request merged**